### DISCUSSION TITLE
`PathValue` and `FieldPathByValue` not working nicely together

### DISCUSSION THREAD

**Discussion by dolma** [react-hook-form/react-hook-form#10419] [Q&A]
Hi all!

I'm having an issue when using the `FieldPathByValue` type along with the `getValues` function.

### How to reproduce:

```ts
interface Props<TFieldValues extends FieldValues> {
    form: UseFormReturn<TFieldValues>;
    name: FieldPathByValue<TFieldValues, string>;
}

const useCustomField = <TFieldValues extends FieldValues>({ form, name }: Props<TFieldValues>) => {
    const { getValues, setValue } = form;

    const value = getValues(name);
    setValue(name, 'New Value');
};
```

### Expected behaviour:

- `value` should be of type `string`
- `setValue` should only allow a string value as its second argument

### Current behaviour:

- `value` is of type `PathValue<TFieldValues, FieldPathByValue<TFieldValues, string>>` which seems to be behaving like `any`.
- `setValue` does not allow a string value to be passed as its second argument. It gives the following error:

```
Argument of type 'string' is not assignable to parameter of
type 'PathValue<TFieldValues, FieldPathByValue<TFieldValues, string>>'.(2345)
```

### Note:

I can get around the `getValues` issue by using a type guard, but the `setValue` issue is a lot more problematic.

### Question:

Am I missing something here or `value` should indeed be of type `string`?
In fact, more generally, `PathValue<TFieldValues, FieldPathByValue<TFieldValues, T>>` should resolve to `T` and not `any`.

Is there any way to get the proper type out of `getValues` and `setValue` or is this just not currently supported in RHF?

**Comment by Faizerz** [comment]
Did you ever find a solution to this?

**Reply by BrendanC23** [reply]
I'm having a similar issue. I want to make a function that takes in a path (that must be of type `number` and the `setValue` function. But the types don't work properly.

```ts
function fun<T extends FieldValues>(
    path: FieldPathByValue<T, number>,
    setValue: UseFormSetValue<T>
) {
    const value = 0;

    setValue(path, value);
}
```

This results in the error `Argument of type 'number' is not assignable to parameter of type 'PathValue<T, FieldPathByValue<T, number>>'`.

**Reply by Shentoza** [reply]
I have the same issue when i want to have a form input that can be reused in multiple components... Could you solve it?

**Reply by briandunn** [reply]
Just want to bump this - really need `PathValue<TFieldValues, FieldPathByValue<TFieldValues, string>>` to be `string`. `FieldPathByValue` is so cool, if only it worked.

**Reply by jmsandiegoo** [reply]
Is there a work around for this?

**Reply by BrendanC23** [reply]
@bluebill1049  Can you take a look at this?

Here's a reproduction in [TypeScript's Playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBDAnmApnA3nAYsFAbAEwAUBDGACwCFEA1EvAVxQBpSK7GUBJcPZ7XIQ5MAzvwCqIlFmggASihgMoAOzgBfOADMoEEHABEUFCQDGMALTkIEANYWtsgwG4AUK60MV54BDWn6PABlRWEUAB4AFTgUAA8YFBUCEQF8AjCRAD4AClc4fO1ZAC44SWlZBSVVKMzmPIKwMnISnDS2Klp6Jij+ERgoYBUAc1rXAEoMevyAemm4AEEoIYYQRPgILQRkNAByPoHhnbhgFJUIeBIREWAhlRIAIzw0GAg4RqgSVYSoOA2t1CmcFmcB27TCPDAeB6qUI7WoYWh+0GI0yOwAdNkAEwAZgALABWMaAxxQEBoqQwMLZRoUfgGB6mAxjNzqVxAA).

**Reply by BrendanC23** [reply]
I created a [new issue](https://github.com/react-hook-form/react-hook-form/issues/12828) about this problem.
